src/compiler/crystal/syntax/lexer.cr

```diff
@@ -1771,78 +1686,54 @@ module Crystal
         if next_char.ascii_number?
           raise "octal constants should be prefixed with 0o"
         else
-          finish_scan_prefixed_number 0_u64, false, start
-        end
-      end
-    end
-
-    def scan_bin_number(start, negative)
-      next_char
-
-      num = 0_u64
-      while true
-        case next_char
-        when '0'
-          num *= 2
-        when '1'
-          num = num * 2 + 1
-        when '_'
-          # Nothing
-        else
-          break
+          first_byte = @reader.string.byte_at(start)
+          @token.type = :NUMBER
+          @token.number_kind = :i32
+          @token.value = case first_byte
+                         when '+' then "+0"
+                         when '-' then "-0"
+                         else          "0"
+                         end
+          set_token_raw_from_start(start)
         end
       end
-
-      finish_scan_prefixed_number num, negative, start
     end
 
-    def scan_octal_number(start, negative)
+    def scan_base_npow2_number(bits_per_character, start, negative)
       next_char
 
-      num = 0_u64
-
-      while true
-        char = next_char
-        if '0' <= char <= '7'
-          num = num * 8 + (char - '0')
-        elsif char == '_'
-        else
-          break
-        end
-      end
-
-      finish_scan_prefixed_number num, negative, start
-    end
-
-    def scan_hex_number(start, negative = false)
-      next_char
+      digits = String::CHAR_TO_DIGIT.to_unsafe
+      digit_count = 0
+      base = 2 ** bits_per_character
+      start_pos = current_pos
 
-      num = 0_u64
       while true
         char = next_char
-        if char == '_'
-        else
-          hex_value = char_to_hex(char) { nil }
-          if hex_value
-            num = num * 16 + hex_value
-          else
-            break
-          end
-        end
-      end
-
-      finish_scan_prefixed_number num, negative, start
-    end
-
-    def finish_scan_prefixed_number(num, negative, start)
-      if negative
-        string_value = (num.to_i64 * -1).to_s
+        next if char == '_'
+        digit = digits[char.ord]
+        break if digit > base || digit < 0
+        digit_count += 1
+      end
+
+      string = string_range(start_pos + 1, current_pos)
+      string_value = nil
+
+      case digit_count * bits_per_character
+      when 0
+        raise "numbers cannot end with a prefix"
+      when 1..32
+        string_value = string.to_u32(base: base, underscore: true).to_s
+      when 33..64
+        string_value = string.to_u64(base: base, underscore: true).to_s
+      when 65..128
+        string_value = string.to_u128(base: base, underscore: true).to_s
       else
-        string_value = num.to_s
+        raise_value_doesnt_fit_in "Int128", string, start
       end
 
+      string_value = string_value.not_nil!
```

Comment:
This `not_nil!` could be avoided if we assign the value of the case to `string_value`.